### PR TITLE
fix(193523): JSDoc optional parameters don't display in functions

### PR DESCRIPTION
**Pull request by perplexyves**
Hello. This PR is a fix to #193523, which was labelled with both `help-wanted` and `bug` labels.

Useful summary of the issue:

- First, go to **settings.json** and add these 2 config lines:
```json
"editor.suggest.preview": true,
"javascript.suggest.completeFunctionCalls": true,
```

- Then, in a JS file type the following:
```js
/**
 * @param {string} one hi.
 * @param {number} two howdy.
 * @param {string} [three] hello.
 */
function test(one, two, three) { }
```

- Now, when you try to invoke the function `test`, you should notice that in the autocompletion preview the optional params have been omitted:

  image

After applying the fix (based on @RyanCavanaugh [advice](https://github.com/microsoft/vscode/issues/193523#issuecomment-1726456324)), the problem is solved.
  
  image

### CODE REVIEW

**Comment by perplexyves** [comment]
Also I am ready to assist with any unit test to cover this case.

**Comment by akobrin** [comment]
I am unfamiliar with this whole review process. Is the bottom line that the fix will be included in a VS Code update in the future? Thank you

**Review by mjbvz — CHANGES_REQUESTED** [review]
We still want to hide optional parameters at the end of the list. Otherwise some library functions will generation an overwhelming number of parameters that you don't care about

[#193523](https://github.com/microsoft/vscode/issues/193523) is specifically about when the optional parameter appears between two non-optional params. Let's focus on that case

Also make sure to add tests. These are defined in `functionCallSnippet.test.ts`

**Comment by perplexyves** [comment]
@microsoft-github-policy-service agree

**Comment by perplexyves** [comment]
I've added the requested changed @mjbvz. Ready to assist if needed.

**Comment by perplexyves** [comment]
Hello Matt. Just fixed the integration tests. Can you please review once available? @mjbvz

File: extensions/typescript-language-features/src/languageFeatures/util/snippetForFunctionCall.ts [hunk 1]
```diff
@@ -75,6 +76,17 @@ function getParameterListParts(
...
 					// Skip this parameter
 					const nameIsThis = part.text === 'this';
+
+					/* Add optional param to temp array. Once a non-optional param is encountered,
+					this means that previous optional params were mid-list ones, thus they should
+					be displayed */
+					if (nameIsFollowedByOptionalIndicator) {
+						optionalParams.push(part);
+					} else {
+						optionalParams.forEach(param => parts.push(param));
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/languageFeatures/util/snippetForFunctionCall.ts:86]:
```suggestion
parts.push(...optionalParams);
```

**Review by mjbvz — APPROVED** [review]
Thanks!

**Review by rzhao271 — APPROVED** [review]

**Comment by perplexyves** [comment]
Pleasure!

**Pull request merged**